### ISSUE TITLE
MaybeNullAttribute not working

### ISSUE THREAD

**Issue by JamesNK** [dotnet/roslyn#36986]
I'm converting Newtonsoft.Json to use nullable types and I can't get `MaybeNullAttribute` to work. My remaining warnings are related to generic types.

Note that I have many old targets so I'm using internal nullable attributes (located in NullableAttributes.cs). `NotNullAttribute` and `NotNullWhenAttribute` work fine, but I can't get `MaybeNullAttribute` to work. I've tried it in `System.Runtime.CompilerServices` and `System.Diagnostics.CodeAnalysis` namespaces with no result.

**Version Used**: VS2019 + Microsoft.Net.Compilers.Toolset=3.2.0-beta3-final

**Steps to Reproduce**:

1. git clone https://github.com/JamesNK/Newtonsoft.Json.git
2. git checkout jamesnk/nullable-types
3. Go to JsonConvert.DeserializeObject<T> method (JsonConvert.cs, line 787)

**Expected Behavior**:

There should not be a nullable warning because `[return: MaybeNull]` is on the return type.

**Actual Behavior**:

There is a warning casting `object?` to `T`.

![image](https://user-images.githubusercontent.com/303201/60654957-0c03a700-9ea1-11e9-9d14-5feda0ca0877.png)

**Comment by RikkiGibson** [comment]
I think this example should not warn:
```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;

namespace System.Diagnostics.CodeAnalysis
{
    class MaybeNullAttribute : System.Attribute { }
}

public class C {
    [return: MaybeNull]
    public T M<T>(object? o) {
        return (T)o; // gives warning CS8601, but probably shouldn't
    }
}
```

**Comment by JamesNK** [comment]
This also shouldn't warn:

```cs
public class C {
    [return: MaybeNull]
    public T M<T>(object? o) {
        return default;
    }
}
```

**Comment by jcouv** [comment]
Thanks for reporting this.
This is currently by-design. The attributes (`[AllowNull]` and such) currently only affect callers. They do not affect implementations or OHI yet. This is tracked by https://github.com/dotnet/roslyn/issues/36039

**Comment by JamesNK** [comment]
What should I do in this situation? Suppress CS8601?

**Comment by jcouv** [comment]
Use the suppression operator. For example: `return default!;`.
